### PR TITLE
[pytorch][build][test]Update Torchserve version in PyTorch Inference 1.12

### PR DESCRIPTION
**Pull request by kenny-ezirim**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Updated Torchserve version in PyTorch Inference images. Required Torchserve pypi dependencies were installed directly from Torchserve requirements source file (common.txt). SM PyTorch Inference toolkit was pinned at 2.0.11 and JDK was upgraded from version 11 to version 17.

### Tests run
Tests executed include sanity, ec2, ecs, and eks tests. [6523062](https://github.com/aws/deep-learning-containers/pull/2745/commits/652306229d0f66438ecb10efbfac3fd5ca9d6079)

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit [ccc7bfa](https://github.com/aws/deep-learning-containers/pull/2745/commits/ccc7bfa8e5326fe0b783a2fc96f9cb571d276c87) for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/inference/docker/1.12/py3/cu116/Dockerfile.gpu [hunk 1]
```diff
@@ -220,10 +219,16 @@ RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pk
 
 # TODO: revert torchserve-nightly back to torchserve once a compatible stable version is released
 RUN pip uninstall -y model-archiver multi-model-server \
- && pip install captum \
  && pip install torchserve==${TS_VERSION} \
  && pip install torch-model-archiver==${TS_VERSION}
 
+# Install TorchServe pypi dependencies directly from their requirements file common.txt
+# NOTE: This also brings in unnecessary cpu dependencies like nvgpu
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/cu116/Dockerfile.gpu:226]: this comment is not relevant for gpu image

**Review by arjkesh — COMMENTED** [review]
We should plan to increment the DLC major version for these containers to "2" with the openjdk major version upgrade. There is a sanity test that makes sure each version is represented, so you may need to create new dockerfiles (i.e. under an openjdk17 directory).

File: pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    opencv-python \
+    packaging \
+    "pyyaml>=5.4,<5.5" \
+ && pip uninstall -y torch torchvision torchaudio torchdata \
+ && pip install --no-cache-dir -U ${PT_INFERENCE_URL} ${PT_TORCHVISION_URL} ${PT_TORCHAUDIO_URL} ${PT_TORCHDATA_URL}
+
+# add necessary certificate for aws sdk cpp download
+RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
+
+# TODO: revert torchserve-nightly back to torchserve once a compatible stable version is released
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu:220]: can you remove this comment? i do not believe we are installing TS nightly here

File: pytorch/inference/docker/1.12/py3/jdk17/Dockerfile.cpu [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+#  | || '_ ` _ \ / _` |/ _` |/ _ \ | |_) / _ \/ __| | '_ \ / _ \
+#  | || | | | | | (_| | (_| |  __/ |  _ <  __/ (__| | |_) |  __/
+# |___|_| |_| |_|\__,_|\__, |\___| |_| \_\___|\___|_| .__/ \___|
+#                      |___/                        |_|
+#################################################################
+
+FROM ec2 AS sagemaker
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/Dockerfile.cpu:223]: this DLC major version should also be set to 2

File: pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+#                                         |___/
+#  ____           _
+# |  _ \ ___  ___(_)_ __   ___ 
+# | |_) / _ \/ __| | '_ \ / _ \
+# |  _ <  __/ (__| | |_) |  __/
+# |_| \_\___|\___|_| .__/ \___|
+#                  |_|
+########################################################
+
+FROM nvidia/cuda:11.3.1-base-ubuntu20.04 AS base_image
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu:25]: I recommend removing the multistage targets here. This is only a sagemaker image, as we do not release the EC2 image from here, so there is no reason for this to be multistage. This will also reduce confusion as you will be able to just have one dlc_major_version defined

File: pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+#  | || '_ ` _ \ / _` |/ _` |/ _ \ | |_) / _ \/ __| | '_ \ / _ \
+#  | || | | | | | (_| | (_| |  __/ |  _ <  __/ (__| | |_) |  __/
+# |___|_| |_| |_|\__,_|\__, |\___| |_| \_\___|\___|_| .__/ \___|
+#                      |___/                        |_|
+#################################################################
+
+FROM ec2 AS sagemaker
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu:282]: I suggest removing the sagemaker recipe from this dockerfile, since we release the SM image as cu113. This will also reduce confusion because of the multiple dlc_major_version labels

File: pytorch/inference/docker/1.12/py3/jdk11/cu113/Dockerfile.gpu
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk11/cu113/Dockerfile.gpu:1]: dlc_major_version in this file looks odd. I suggest making this a single dockerfile without multistage targets, and just include one dlc_major_version that corresponds to the sagemaker image, since we don't release the ec2-image version of this dockerfile anyway

File: pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    opencv-python \
+    packaging \
+    "pyyaml>=5.4,<5.5" \
+ && pip uninstall -y torch torchvision torchaudio torchdata \
+ && pip install --no-cache-dir -U ${PT_INFERENCE_URL} ${PT_TORCHVISION_URL} ${PT_TORCHAUDIO_URL} ${PT_TORCHDATA_URL}
+
+# add necessary certificate for aws sdk cpp download
+RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
+
+# TODO: revert torchserve-nightly back to torchserve once a compatible stable version is released
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu:220]: This comment can be removed

File: pytorch/inference/buildspec-1-12.yml [hunk 7]
```diff
@@ -34,10 +34,12 @@ images:
     image_size_baseline: 4899
     device_type: &DEVICE_TYPE cpu
     python_version: &DOCKER_PYTHON_VERSION py3
+    jdk_version: &DOCKER_JDK_VERSION jdk17
     tag_python_version: &TAG_PYTHON_VERSION py38
+    tag_jdk_version: &TAG_JDK_VERSION jdk17
     os_version: &OS_VERSION ubuntu20.04
-    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-ec2" ]
-    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile., *DEVICE_TYPE ]
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *TAG_JDK_VERSION, "-", *OS_VERSION, "-ec2" ]
```
Comment (arjkesh) [inline pytorch/inference/buildspec-1-12.yml:41]: Please remove jdk version from tag - this will have an impact on release process and isn't something we can easily introduce

File: pytorch/inference/buildspec-1-12.yml [hunk 8]
```diff
@@ -47,11 +49,13 @@ images:
...
     device_type: &DEVICE_TYPE gpu
     python_version: &DOCKER_PYTHON_VERSION py3
+    jdk_version: &DOCKER_JDK_VERSION jdk17
     tag_python_version: &TAG_PYTHON_VERSION py38
+    tag_jdk_version: &TAG_JDK_VERSION jdk17
     cuda_version: &CUDA_VERSION cu116
     os_version: &OS_VERSION ubuntu20.04
-    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION, "-ec2" ]
-    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile.,
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *TAG_JDK_VERSION, "-",  *CUDA_VERSION, "-", *OS_VERSION, "-ec2" ]
```
Comment (arjkesh) [inline pytorch/inference/buildspec-1-12.yml:57]: Please remove jdk version from tag - this will have an impact on release process and isn't something we can easily introduce

File: pytorch/inference/buildspec-1-12.yml [hunk 9]
```diff
@@ -75,11 +81,13 @@ images:
...
     device_type: &DEVICE_TYPE gpu
     python_version: &DOCKER_PYTHON_VERSION py3
+    jdk_version: &DOCKER_JDK_VERSION jdk17
     tag_python_version: &TAG_PYTHON_VERSION py38
+    tag_jdk_version: &TAG_JDK_VERSION jdk17
     cuda_version: &CUDA_VERSION cu113
     os_version: &OS_VERSION ubuntu20.04
-    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION, "-sagemaker" ]
-    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /, *CUDA_VERSION, /Dockerfile.,
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *TAG_JDK_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION, "-sagemaker" ]
```
Comment (arjkesh) [inline pytorch/inference/buildspec-1-12.yml:89]: Please remove jdk version from tag - this will have an impact on release process and isn't something we can easily introduce

File: test/test_utils/__init__.py [hunk 10]
```diff
@@ -248,6 +254,13 @@ def get_dockerfile_path_for_image(image_uri):
         if "example" not in path
     ]
 
+    if not jdk_version and any("jdk" in path for path in dockerfiles_list):
+        raise LookupError(
```
Comment (arjkesh) [inline test/test_utils/__init__.py:258]: this needs to be updated - jdk version should not be in the image URI - does dockerfiles_list not pick it up? it should as it should look at subdirs

File: test/test_utils/__init__.py [hunk 11]
```diff
@@ -235,6 +235,12 @@ def get_dockerfile_path_for_image(image_uri):
     if not os.path.isdir(python_version_path):
         python_version_path = os.path.join(framework_version_path, "py3")
 
+    jdk_version = re.search(r"jdk\d+", image_uri)
```
Comment (arjkesh) [inline test/test_utils/__init__.py:238]: this needs to be updated - jdk version should not be in the image URI

File: pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+#              |___/
+#  ___                              ____           _
+# |_ _|_ __ ___   __ _  __ _  ___  |  _ \ ___  ___(_)_ __   ___
+#  | || '_ ` _ \ / _` |/ _` |/ _ \ | |_) / _ \/ __| | '_ \ / _ \
+#  | || | | | | | (_| | (_| |  __/ |  _ <  __/ (__| | |_) |  __/
+# |___|_| |_| |_|\__,_|\__, |\___| |_| \_\___|\___|_| .__/ \___|
+#                      |___/                        |_|
+#################################################################
+
+FROM nvidia/cuda:11.3.1-base-ubuntu20.04
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu:25]: change to `FROM nvidia/cuda:11.3.1-base-ubuntu20.04 AS sagemaker`

File: pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+#                                         |___/
+#  ____           _
+# |  _ \ ___  ___(_)_ __   ___ 
+# | |_) / _ \/ __| | '_ \ / _ \
+# |  _ <  __/ (__| | |_) |  __/
+# |_| \_\___|\___|_| .__/ \___|
+#                  |_|
+########################################################
+
+FROM nvidia/cuda:11.6.2-base-ubuntu20.04
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu:25]: change to `FROM nvidia/cuda:11.6.2-base-ubuntu20.04 AS ec2`

File: pytorch/inference/docker/1.12/py3/jdk11/cu113/Dockerfile.gpu
Comment (arjkesh) [reply pytorch/inference/docker/1.12/py3/jdk11/cu113/Dockerfile.gpu:1]: this still remains an issue - you'll have inconsistent dockerfile versioning if you do not update this file

File: pytorch/inference/docker/1.12/py3/jdk11/cu116/Dockerfile.gpu
[the remark this answers is not in the document]
Comment (arjkesh) [reply pytorch/inference/docker/1.12/py3/jdk11/cu116/Dockerfile.gpu:1]: this still remains an issue - you'll have inconsistent dockerfile versioning if you do not update this file

File: pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+    libgomp1 \
+    libibverbs-dev \
+    libnuma1 \
+    libnuma-dev \
+    libsm6 \
+    libssl1.1 \
+    libssl-dev \
+    libxext6 \
+    libxrender-dev \
+    openjdk-17-jdk \
```
Comment (kenny-ezirim) [inline pytorch/inference/docker/1.12/py3/jdk17/cu116/Dockerfile.gpu:107]: please review approve the adoption of JDK17+TS 0.6.1

File: pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    libgomp1 \
+    libibverbs-dev \
+    libnuma1 \
+    libnuma-dev \
+    libsm6 \
+    libssl1.1 \
+    libssl-dev \
+    libxext6 \
+    libxrender-dev \
+    openjdk-17-jdk \
```
Comment (kenny-ezirim) [inline pytorch/inference/docker/1.12/py3/jdk17/cu113/Dockerfile.gpu:108]: please review approve the adoption of JDK17+TS 0.6.1
